services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java

```diff
@@ -1114,7 +1117,7 @@ protected Response authenticationComplete() {
         if (nextRequiredAction != null) {
             return AuthenticationManager.redirectToRequiredActions(session, realm, authenticationSession, uriInfo, nextRequiredAction);
         } else {
-            event.detail(Details.CODE_ID, authenticationSession.getParentSession().getId());  // todo This should be set elsewhere.  find out why tests fail.  Don't know where this is supposed to be set
+            event.detail(Details.CODE_ID, authenticationSession.getAuthNote(USER_SESSION_ID) != null ? authenticationSession.getAuthNote(USER_SESSION_ID) : authenticationSession.getParentSession().getId());  // todo This should be set elsewhere.  find out why tests fail.  Don't know where this is supposed to be set
```

Comment:
I see this code snippet in several places:

```
authenticationSession.getAuthNote(USER_SESSION_ID) != null ? authenticationSession.getAuthNote(USER_SESSION_ID) : authenticationSession.getParentSession().getId()
```

I fail to understand the intent of that check. Could that be refactored as a method of maybe `authenticationSession` with a meaningful name and a code comment when on or the other is taken? 

My first guess is that old authentication sessions might not have it, so is this change about a zero-downtime upgrades? I'd assume that those would be short lived, and that those extra could would not be needed...

Reply:
So now we write sometimes the user session id in the event, sometimes the parent session id. I don't understand in which case one or the other happens by looking at the local code snippets or the whole change. 

I'd like to see a good definition of what the field in the event should contain and a use case how it should be used. If we decide to break an existing definition, we can put this in the release notes. For example: If we decide it should always contain the user session id, then it should always contain the user session, and null if there is no user session for some reason. Looking at the name of the field `code_id` I'd usually expect something different.

A similar pattern that tenary statement exists not only for events, but also in the AuthenticatinoProcessor; again some comment why this is the way would help me and I'd assume future readers of the code as well.

```
            String userSessionId = authSession.getAuthNote(USER_SESSION_ID);
            userSession = lockUserSessionsForModification(session,
                    () -> session.sessions().getUserSession(realm, userSessionId != null ? userSessionId : authSession.getParentSession().getId()));
```